zebra-chain/src/primitives/zcash_history.rs

```diff
@@ -0,0 +1,329 @@
+        }
+        Ok(new_nodes)
+    }
+
+    /// Remove the last leaf (block) from the tree.
+    ///
+    /// Returns the number of nodes removed from the tree after the operation.
+    fn truncate_leaf(&mut self) -> Result<u32, zcash_history::Error> {
+        self.tree.truncate_leaf()
+    }
+
+    /// Return the root hash of the tree, i.e. `hashChainHistoryRoot`.
+    fn hash(&self) -> ChainHistoryMmrRootHash {
+        // Both append_leaf() and truncate_leaf() leave a root node, so it should
+        // always exist.
+        self.tree
+            .root_node()
+            .expect("must have root node")
+            .data()
+            .hash()
+            .into()
+    }
+}
+
+/// Convert a Block into a zcash_history::NodeData used in the MMR tree.
+///
+/// `sapling_root` is the root of the Sapling note commitment tree of the block.
+fn block_to_history_node(
+    block: Arc<Block>,
+    network: Network,
+    sapling_root: &sapling::tree::Root,
+) -> zcash_history::NodeData {
+    let height = block
+        .coinbase_height()
+        .expect("block must have coinbase height during contextual verification");
+    let branch_id = ConsensusBranchId::current(network, height)
+        .expect("must have branch ID for chain history network upgrades");
+    let block_hash = block.hash().0;
+    let time: u32 = block
+        .header
+        .time
+        .timestamp()
+        .try_into()
+        .expect("deserialized and generated timestamps are u32 values");
+    let target = block.header.difficulty_threshold.0;
+    let sapling_root: [u8; 32] = sapling_root.into();
+    let work = block
+        .header
+        .difficulty_threshold
+        .to_work()
+        .expect("work must be valid during contextual verification");
+    // There is no direct `std::primitive::u128` to `bigint::U256` conversion
+    let work = bigint::U256::from_big_endian(&work.as_u128().to_be_bytes());
+
+    let sapling_tx_count = count_sapling_transactions(block);
+
+    zcash_history::NodeData {
+        consensus_branch_id: branch_id.into(),
+        subtree_commitment: block_hash,
+        start_time: time,
+        end_time: time,
+        start_target: target,
+        end_target: target,
+        start_sapling_root: sapling_root,
+        end_sapling_root: sapling_root,
+        subtree_total_work: work,
+        start_height: height.0 as u64,
+        end_height: height.0 as u64,
+        sapling_tx: sapling_tx_count,
+    }
+}
+
+/// Count how many Sapling transactions exist in a block,
+/// i.e. transactions "where either of vSpendsSapling or vOutputsSapling is non-empty"
+/// (https://zips.z.cash/zip-0221#tree-node-specification).
+fn count_sapling_transactions(block: Arc<Block>) -> u64 {
+    block
+        .transactions
+        .iter()
+        .filter(|tx| tx.has_sapling_shielded_data())
+        .count()
+        .try_into()
+        .expect("number of transactions must fit u64")
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Can you put the test module in own file?

We're moving towards a standard module structure, these tests should go in `zcash_history::tests::vectors`.

Reply:
Done!